File: src/Functional/OCaml/factory_method.ml
type database = {
  connect : unit -> unit;
  query : unit -> unit;
}

let create_postgres () = {
  connect = (fun () -> print_endline "PostgreSQL connect");
  query = (fun () -> print_endline "PostgreSQL query");
}

let create_mysql () = {
  connect = (fun () -> print_endline "MySQL connect");
  query = (fun () -> print_endline "MySQL query");
}

let use_database create_database =
  let database = create_database () in
  database.connect ();
  database.query ()

let () =
  use_database create_postgres;
  use_database create_mysql

File: src/Functional/Haskell/FactoryMethod.hs
data Database = Database
  { connect :: IO ()
  , query :: IO ()
  }

createPostgres :: IO Database
createPostgres = pure Database
  { connect = putStrLn "PostgreSQL connect"
  , query = putStrLn "PostgreSQL query"
  }

createMySql :: IO Database
createMySql = pure Database
  { connect = putStrLn "MySQL connect"
  , query = putStrLn "MySQL query"
  }

useDatabase :: IO Database -> IO ()
useDatabase createDatabase = do
  database <- createDatabase
  connect database
  query database

main :: IO ()
main = do
  useDatabase createPostgres
  useDatabase createMySql
